File: python/Problem2.py
# Problem: 2. Add Two Numbers
# URL: https://leetcode.com/problems/add-two-numbers/
# Difficulty: Medium
# Tags: Math, Linked List

from typing import Optional

# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next

class Solution:
  def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    head: Optional[ListNode] = ListNode(0)
    current: Optional[ListNode] = head
    carry: int = 0

    while l1 is not None or l2 is not None or carry != 0:
      d1: int = l1.val if l1 is not None else 0
      d2: int = l2.val if l2 is not None else 0

      sum: int = d1 + d2 + carry
      d: int = sum % 10
      carry: int = sum // 10

      current.next = ListNode(d)
      current = current.next

      l1 = l1.next if l1 is not None else None
      l2 = l2.next if l2 is not None else None
    
    result: Optional[ListNode] = head.next
    head.next = None
    return result

File: cpp/Problem2.cpp
/*
 * Problem: 2. Add Two Numbers
 * 
 * Difficulty: Medium
 * Tags: Math, Linked List
 */

using namespace std;

// Definition for singly-linked list.
struct ListNode {
	int val;
	ListNode *next;
	ListNode() : val(0), next(nullptr) {}
	ListNode(int x) : val(x), next(nullptr) {}
	ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
	ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
		ListNode* head = new ListNode();
		ListNode* current = head;
		int carry = 0;

		while (l1 != nullptr || l2 != nullptr || carry != 0) {
			int d1 = l1 != nullptr ? l1->val : 0;
			int d2 = l2 != nullptr ? l2->val : 0;

			int sum = d1 + d2 + carry;
			int d = sum % 10;
			carry = sum / 10;

			current->next = new ListNode(d);
			current = current->next;

			l1 = l1 != nullptr ? l1->next : nullptr;
			l2 = l2 != nullptr ? l2->next : nullptr;
		}
		
		ListNode* result = head->next;
		delete head;
		return result;
	}
};